### ISSUE TITLE
When an Entity is deleted, warn the user if it is referenced by a script in the scene

### ISSUE THREAD

**Issue by yaustar** [playcanvas/editor#287]
This is useful for teams where it's difficult to know if an Entity is referenced elsewhere in the scene. When deleting the Entity, we could warn the user that it is being referenced by other Entities and give them a list of paths to where they are in the scene hierarchy.

A stretch goal would be to find all references to an Entity in a similar way that we do with assets.

**Comment by Maksims** [comment]
I would argue about "notification" approach (it is obtrusive), and follow same approach it is with assets: we have "References" context menu, which shows where it is referenced. So developer can make a decision himself, check references or remove without concerns.

"Replace" context menu on assets is really usefull too, but I'm afraid it doesn't make direct sense on entity, perhaps different wording. Would be a good Editor Plugin too ;)

A better way for showing missing references (entities and assets), would be a good feature too. There are a lot of cases where missing references can be a thing, like copying entities/assets, removing stuff, etc.

**Comment by yaustar** [comment]
Would that be that intrusive if it only happened when it is being referenced elsewhere though? 

I would rather know when it happens rather than manually checking every time I consider deleting an entity. Is it often that someone would delete an Entity and not want to know if doing so could potentially break another Entity's logic?

**Comment by Maksims** [comment]
> Would that be that intrusive if it only happened when it is being referenced elsewhere though?

Any conditional intrusion - is bad. Especially when it is conditional. Assets deletion for example is not conditional, and is really important save measure, as undoing assets deletion is not possible.

> Is it often that someone would delete an Entity and not want to know if doing so could potentially break another Entity's logic?

I often do so, or am aware that it is refernced, so will delete something, than update reference for example.

What better would be, is a way to show that some entity requires an attention for example, subtle but visible in hierarchy. That attention can be based on missing mandatory attribute as suggested in https://github.com/playcanvas/editor/issues/288 or a missing reference. Same would apply to asset, like a missing reference to texture.

Subtle consistent icon, that could be used across everywhere - would be a good and non intrusive way to help developers.
And missing references - is a valid state for application. Usually it is best to make sure your scripts take it in an account.

**Comment by yaustar** [comment]
> I often do so, or am aware that it is refernced, so will delete something, than update reference for example.

In a team scenario, its harder to be aware that it is being referenced. We could add an option to enable/disable this warning in the user settings so it's up to the user? 

Having an icon would be great but it's something that can be missed if they aren't looking at the hierarchy and instead, more focused in the 3D view.

**Comment by Maksims** [comment]
Intrusive popup, will only help to inform of deletion. It does not help to re-assign references, nor it helps to do anything after the deletion happened.
Popups - is in UX is almost always a bad UX.

To solve an issue of missing references, assist developer of finding where references are missing, and provide a ways to change references, we need:

1. "References" context menu on Entity, same as on Asset.
2. Subtle icon "call for attention", on entities (in hierarchy) and assets (assets panel), to show missing references and/or mandatory attributes (discussed new feature in issue mentioned before).

**Comment by Maksims** [comment]
> but doesn't help when they need to know what it used to reference to and how to get it back.

Indeed, nor the popup helps them to identify it after referenced entity was deleted. This will go even worse with deleting hierarchies of entities. So intrusive popup does very little help. And no help at all in a collaborative context.

**Comment by yaustar** [comment]
> Indeed, nor the popup helps them to identify it after referenced entity was deleted

It will pop up before it gets deleted though.

**Comment by Maksims** [comment]
> It will pop up before it gets deleted though.

I understand a proposed UX, but again, this gives no use to after the deletion, or in teams it can also be not known to person who is deleting what to do with references.

Perhaps, being a popups intrusive, if there would be a way to mark a script attribute as "mandatory", then it would show a popup (unless ever marked to "do not show").

**And still, it is only helps at the moment before deletion. While there is no way to find out before deletion (no Reference context menu), and does not help after deletion (no ways to know where mandatory references are missing).**

**Comment by yaustar** [comment]
Oh yeah, I'm not saying that this is an either/or situation. Personally, I would like to have all these options. Popup warning, indicators in the scene hierarchy and in the inspector too.

> Perhaps, being a popups intrusive, if there would be a way to mark a script attribute as "mandatory"

I was wondering about this too. Could definitely start with that and see if the indicators would be enough on its own?

**Comment by Maksims** [comment]
> Could definitely start with that and see if the indicators would be enough on its own?

It definitely will help, but there is never enough 😆
